Report detected artifacts with relative XYZ offset

m3d opened the pull request:
The artifacts are now reported with offset (in millimetres), X forward, Y left, Z up. This PR currently breaks real robots using old color based classifier. The `artf` data format changed from `[type, angle, distance]` to `[type, [rel_x, rel_y, rel_z]]`.

In order to compute distances `fx` has to be specified in config file (at the moment SSCI X4 config2 drone is used as default).

There are two new streams `debug_result` and `debug_cv_result` where is string (I failed to properly serialize the numpy data) containing data of each classifier. It is possible to replay position computation based on these two streams.

Camera offset/rotation is ignored here and must be taken into account when calculating precise global artifact position.

m3d commented:
```
closest_artifact_report:
  name: backpack_3
  type: TYPE_BACKPACK
  true_pos: 80.16 -4.87234 0.32668
  reported_pos: 79.7 -4.85 0.41
  distance: 0.4680184369392776 
```
i.e. better 👍 
ver82rc2s1 1833543f-9f25-4f38-98c6-9eb7d141318e

zwn commented on subt/main.py:
```diff
@@ -545,16 +547,17 @@ def on_acc(self, timestamp, data):
         x0, y0, z0 = self.offset
-        angle, dist = self.yaw + math.radians(deg_100th / 100.0), dist_mm / 1000.0
-        ax = x0 + x + math.cos(angle) * dist
-        ay = y0 + y + math.sin(angle) * dist
-        az = z0 + z
+        ax = x0 + x + dx/1000.0
+        ay = y0 + y + dy/1000.0
+        az = z0 + z + dz/1000.0
```
`self.xyz` is set in `on_pose3d` and I believe it is already corrected for origin offset (it either comes from `odoimuloc` or from `offseter`).

m3d replied:
Note, that I did not change this part - and offseter is not plugged in as input ... only "independent reference" (as discussed over email). I.e. so far I keep the "compatibility" here.

zwn replied:
self.xyz used to come only from pose2d which has always been without origin offset
pose3d was from the begining with origin offset but was not used for anything inside app
somewhere along the line on_pose3d started to set self.xyz as well at which point it become mangled and beyond my capacity to comprehend

zwn commented on subt/main.py:
```diff
@@ -545,16 +547,17 @@ def on_acc(self, timestamp, data):
             # there can be observed artifact (false) on the start before the coordinate system is defined
             return
-        artifact_data, deg_100th, dist_mm = data
+        artifact_data, dx, dy, dz = data
+        vector = [dx, dy, dz]
```
Can the data come as [`artifact_data`, [dx, dy, dz]]? So that there are 2 components - first is artifact type and second is 3d vector of relative position.

m3d replied:
Good point - now is the good moment to do it - the data are not backward compatible here. I used the same types as output `artf_xyz` which is the only discouraging detail (but there is list of reports, i.e. it is not 1:1).

m3d replied:
I change that here, but not the rest of all `artf_xyz` reports - please add it to the dept list, thanks.

zwn commented on subt/test_artf_node.py:
```diff
@@ -58,4 +60,53 @@ def test_merge_two_results(self):
         self.assertEqual(checked_result, expected_result)
 
 
+if __name__ == "__main__":
```
So far we have been using `test_<something>.py` to only run unittests. This kind of tests were usually in the `<something>.py` file. I am not saying we cannot change the convention, just that we should do it consciously.

m3d replied:
well this was supposed to be "extended test" (regression test?) based on previously logged that, but in meantime I found out that I do not have necessary inputs to do that (and the logfiles twice changed recently). I may move it, thanks.

m3d replied:
OK, moved but ignore it at the moment - I need to log the detector outputs stream so that I can analyze the "fusion" and then the depth/xyz calculation

m3d commented:
I am running simple test S1 - 9b56a8e4-a5a9-4493-8277-2c6c6462fda6

m3d commented:
`ver82rc4` is still crashing on CloudSim with the last data in log
```
0:00:05.371840 26 b"stdout [('backpack', [(624, 20, 0.97964704), (628, 20, 0.98307204), (624, 24, 0.97980016), (628, 24, 0.96485513)])] []"
0:00:05.372017 1 None
0:00:05.372136 15 [0.11938408017158508]
0:00:05.374577 27 [b'A150L', -5.999997, 5.000044, 0.146687, -0.000169, 2.8e-05, -0.037868, 0.999283]
```
I am going to rollback storing separately results of MDNET, as this is the next following call. The structure is probably more complex, so `str()` may work.

m3d commented:
`ver82rc5` - the drone move a little bit inside and crahsed again - df37dff6-7c0b-48ed-bbb1-1c1ec8fab406

m3d commented:
it seems to be working again (`ver82rc6`) if I abandon logging of detector results in "native" format - running stafeta and 3freyjas in P1 (when ready)

zwn commented:
The problem was related to this:
```
TypeError: can not serialize 'numpy.int64' object
```
It is listed in the console when run locally. 

To wholly convert numpy array to python list one must call https://numpy.org/doc/stable/reference/generated/numpy.ndarray.tolist.html which also numpy scalars to python scalars.

jisa commented on subt/artf_node.py:
```diff
@@ -133,13 +162,65 @@ def detect(self, image):
+    names = lookup_stream_names(args.logfile)
+    stdout_stream_id = lookup_stream_id(args.logfile, 'detector.stdout')
+    streams = [stdout_stream_id]
+    if 'detector.debug_depth' in names:
```
Why is this a hard-coded name and not a commandline parameter?

jisa commented on subt/artf_node.py:
```diff
@@ -133,13 +162,65 @@ def detect(self, image):
+    else:
+        depth_stream_id = None
+
+    # dummy depth for loder logfiles
```
s/loder/older ?

m3d replied:
@jisa thanks for comments/review - this part is still "draft" and I was thinking to drop it (as I was unable to debug the logging - I am no longer able to run simulation "reasonably", and loop via CloudSim with broken logging is not nice either :( 
Now my plan is to support only new logs and ignore old ones, i.e. this section I wil rework now.

jisa commented on subt/artf_node.py:
```diff
@@ -133,13 +162,65 @@ def detect(self, image):
+    # dummy depth for loder logfiles
+    row = [5000] * 640
+    depth = np.array([row] * 360, dtype=np.uint16)
+    fx = 554.25469
```
Another candidate for a commandline parameter?

m3d replied:
new logs have it in config, so I will read it from there

m3d replied:
it should be fixed now. Can be tested on Freyja logfile via
```
python subt\artf_node.py logs\aws\S1\ver82rc6cs1r2\A\aws-ver82rc6cs1r2-A150L.log
```

jisa commented on subt/artf_node.py:
```diff
@@ -133,13 +162,65 @@ def detect(self, image):
+                   only_stream_id=streams) as logreader:
+        for time, stream, msg_data in logreader:
+            if args.time_limit_sec is not None:
+                if time.total_seconds() > args.time_limit_sec:
```
The two conditions above could be merged into one.

m3d replied:
OK, simplified

jisa commented on subt/main.py:
```diff
@@ -497,6 +498,7 @@ def on_pose3d(self, timestamp, data):
             # we cannot align global coordinates if offset is not known
             return
         xyz, rot = data
+        self.orientation = rot  # quaternion
```
What is the plan for robots/configurations that still use on_pose2d?

m3d replied:
it is even worse - what to do with robots, which use old artifact reporter -> it will crash now. I suggest to add it to #640 and review it a week later

m3d commented:
@jisa @zwn I updated the description of this PR and removed "Draft" label. I would proceed - at the moment I am not sure if to keep `result2list()` and `cv_result2list()` functions, which did not work (the data is probably some numpy or torch structure?). I would like to change it but as I am not able to debug it locally it is really pain ... so I would leave it now, and change focus to next task, OK?

zwn commented:
The problem is/was this:

> The problem was related to this:
> 
> ```
> TypeError: can not serialize 'numpy.int64' object
> ```
> 
> It is listed in the console when run locally.
> 
> To wholly convert numpy array to python list one must call https://numpy.org/doc/stable/reference/generated/numpy.ndarray.tolist.html which also numpy scalars to python scalars.

m3d commented:
@zwn - thanks. I am going to update #671 and create image `ver82rc7` from it.

m3d commented:
(and then I will fix `__main__` part for newly collected log)
